File: backend/src/bot/linkedin-auth.js
// backend/src/bot/linkedin-auth.js
// Handles LinkedIn login with session persistence to avoid repeated logins

const { chromium } = require('playwright');
const { humanType, humanClick, humanDelay, getBrowserConfig, sleep } = require('../utils/humanize');
const { accounts } = require('../db');

const LINKEDIN_LOGIN_URL = 'https://www.linkedin.com/login';
const LINKEDIN_FEED_URL  = 'https://www.linkedin.com/feed';

/**
 * Launch browser with optional saved session
 */
const launchBrowser = async (sessionData = null) => {
  const config = getBrowserConfig();
  const browser = await chromium.launch(config);
  
  const contextOptions = {
    viewport: config.viewport,
    userAgent: config.args.find(a => a.startsWith('--user-agent'))?.split('=').slice(1).join('='),
    locale: 'en-US',
    timezoneId: 'America/New_York',
    // Anti-detection: mask webdriver flag
    extraHTTPHeaders: { 'Accept-Language': 'en-US,en;q=0.9' },
  };

  if (sessionData) {
    contextOptions.storageState = sessionData;
  }

  const context = await browser.newContext(contextOptions);
  
  // Mask automation flags
  await context.addInitScript(() => {
    Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
    window.chrome = { runtime: {} };
    Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3] });
    Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
  });

  return { browser, context };
};

/**
 * Check if already logged in
 */
const isLoggedIn = async (page) => {
  try {
    await page.goto(LINKEDIN_FEED_URL, { waitUntil: 'domcontentloaded', timeout: 15000 });
    await sleep(2000);
    const url = page.url();
    return url.includes('/feed') || url.includes('/mynetwork');
  } catch {
    return false;
  }
};

/**
 * Perform LinkedIn login
 */
const login = async (page, email, password, logger) => {
  logger.info(`Logging in as ${email}`);

  await page.goto(LINKEDIN_LOGIN_URL, { waitUntil: 'domcontentloaded', timeout: 20000 });
  await sleep(2000);

  // Already redirected to feed?
  if (page.url().includes('/feed')) {
    logger.info('Already logged in via session');
    return true;
  }

  try {
    await humanType(page, '#username', email, { clear: true, speed: 'normal' });
    await humanDelay('fast');
    await humanType(page, '#password', password, { clear: true, speed: 'normal' });
    await humanDelay('fast');
    await humanClick(page, '[data-litms-control-urn="login-submit"]');

    // Wait for redirect
    await page.waitForURL((url) =>
      !url.toString().includes('/login') && !url.toString().includes('/checkpoint'),
      { timeout: 30000 }
    );

    await sleep(3000);

    // Handle CAPTCHA or verification challenges
    const currentUrl = page.url();
    if (currentUrl.includes('/checkpoint') || currentUrl.includes('/challenge')) {
      logger.warn('Security checkpoint detected — manual intervention may be required');
      // Wait up to 60s for user to solve CAPTCHA manually
      await page.waitForURL((url) => !url.toString().includes('/checkpoint'), { timeout: 60000 });
    }

    if (page.url().includes('/feed') || page.url().includes('/mynetwork')) {
      logger.info('Login successful');
      return true;
    }

    logger.error('Login failed — unexpected URL: ' + page.url());
    return false;

  } catch (err) {
    logger.error('Login error', { error: err.message });
    return false;
  }
};

/**
 * Create authenticated browser session for an account
 * Reuses saved session if valid, otherwise performs fresh login
 */
const createSession = async (account, logger) => {
  const savedSession = await accounts.getSession(account.id);
  
  const { browser, context } = await launchBrowser(savedSession);
  const page = await context.newPage();

  let loggedIn = false;

  if (savedSession) {
    loggedIn = await isLoggedIn(page);
    if (loggedIn) {
      logger.info('Session restored from saved state');
    } else {
      logger.info('Saved session expired, performing fresh login');
    }
  }

  if (!loggedIn) {
    loggedIn = await login(page, account.email, account.password, logger);
    if (loggedIn) {
      // Save session for future use
      const state = await context.storageState();
      await accounts.saveSession(account.id, state);
      logger.info('Session saved to database');
    }
  }

  return { browser, context, page, loggedIn };
};

module.exports = { createSession, login, isLoggedIn, launchBrowser };

File: backend/app/bot/linkedin_auth.py
from __future__ import annotations

from typing import Any

from playwright.sync_api import sync_playwright

from ..db import accounts
from ..utils.humanize import get_browser_config, human_click, human_delay, human_type, sleep


LINKEDIN_LOGIN_URL = "https://www.linkedin.com/login"
LINKEDIN_FEED_URL = "https://www.linkedin.com/feed"


def launch_browser(session_data: dict[str, Any] | None = None) -> dict[str, Any]:
    config = get_browser_config()
    playwright = sync_playwright().start()
    browser = playwright.chromium.launch(headless=config["headless"], args=config["args"])

    user_agent = next(
        (
            arg.split("=", 1)[1]
            for arg in config["args"]
            if arg.startswith("--user-agent=") and "=" in arg
        ),
        None,
    )

    context_options: dict[str, Any] = {
        "viewport": config["viewport"],
        "locale": "en-US",
        "timezone_id": "America/New_York",
        "extra_http_headers": {"Accept-Language": "en-US,en;q=0.9"},
    }
    if user_agent:
        context_options["user_agent"] = user_agent
    if session_data:
        context_options["storage_state"] = session_data

    context = browser.new_context(**context_options)
    context.add_init_script(
        """
        Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
        window.chrome = { runtime: {} };
        Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3] });
        Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
        """
    )

    return {"playwright": playwright, "browser": browser, "context": context}


def is_logged_in(page: Any) -> bool:
    try:
        page.goto(LINKEDIN_FEED_URL, wait_until="domcontentloaded", timeout=15000)
        sleep(2000)
        url = page.url
        return "/feed" in url or "/mynetwork" in url
    except Exception:
        return False


def login(page: Any, email: str, password: str, logger: Any) -> bool:
    logger.info(f"Logging in as {email}")

    page.goto(LINKEDIN_LOGIN_URL, wait_until="domcontentloaded", timeout=20000)
    sleep(2000)

    if "/feed" in page.url:
        logger.info("Already logged in via session")
        return True

    try:
        human_type(page, "#username", email, clear=True, speed="normal")
        human_delay("fast")
        human_type(page, "#password", password, clear=True, speed="normal")
        human_delay("fast")
        human_click(page, '[data-litms-control-urn="login-submit"]')

        page.wait_for_url(
            lambda url: "/login" not in str(url) and "/checkpoint" not in str(url),
            timeout=30000,
        )
        sleep(3000)

        current_url = page.url
        if "/checkpoint" in current_url or "/challenge" in current_url:
            logger.warn("Security checkpoint detected - manual intervention may be required")
            page.wait_for_url(lambda url: "/checkpoint" not in str(url), timeout=60000)

        if "/feed" in page.url or "/mynetwork" in page.url:
            logger.info("Login successful")
            return True

        logger.error(f"Login failed - unexpected URL: {page.url}")
        return False
    except Exception as err:
        logger.error("Login error", {"error": str(err)})
        return False


def create_session(account: dict[str, Any], logger: Any) -> dict[str, Any]:
    saved_session = accounts.get_session(account["id"])

    session = launch_browser(saved_session)
    context = session["context"]
    page = context.new_page()

    logged_in = False

    if saved_session:
        logged_in = is_logged_in(page)
        if logged_in:
            logger.info("Session restored from saved state")
        else:
            logger.info("Saved session expired, performing fresh login")

    if not logged_in:
        logged_in = login(page, account["email"], account["password"], logger)
        if logged_in:
            state = context.storage_state()
            accounts.save_session(account["id"], state)
            logger.info("Session saved to database")

    session["page"] = page
    session["logged_in"] = logged_in
    return session

